### PR TITLE
Specialize `StepBy<Range<{integer}>>`

### PR DESCRIPTION
**Pull request by the8472**
OLD

    iter::bench_range_step_by_fold_u16      700.00ns/iter +/- 10.00ns
    iter::bench_range_step_by_fold_usize    519.00ns/iter  +/- 6.00ns
    iter::bench_range_step_by_loop_u32      555.00ns/iter  +/- 7.00ns
    iter::bench_range_step_by_sum_reducible  37.00ns/iter  +/- 0.00ns

NEW

    iter::bench_range_step_by_fold_u16       49.00ns/iter +/- 0.00ns
    iter::bench_range_step_by_fold_usize    194.00ns/iter +/- 1.00ns
    iter::bench_range_step_by_loop_u32       98.00ns/iter +/- 0.00ns
    iter::bench_range_step_by_sum_reducible   1.00ns/iter +/- 0.00ns

NEW + `-Ctarget-cpu=x86-64-v3`

    iter::bench_range_step_by_fold_u16      22.00ns/iter +/- 0.00ns
    iter::bench_range_step_by_fold_usize    80.00ns/iter +/- 1.00ns
    iter::bench_range_step_by_loop_u32      41.00ns/iter +/- 0.00ns
    iter::bench_range_step_by_sum_reducible  1.00ns/iter +/- 0.00ns

I have only optimized for walltime of those methods, I haven't tested whether it eliminates bounds checks when indexing into slices via things like `(0..slice.len()).step_by(16)`.

### CODE REVIEW

File: library/core/src/iter/adapters/step_by.rs [hunk 1]
```diff
@@ -32,12 +36,12 @@ where
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        if self.first_take {
-            self.first_take = false;
-            self.iter.next()
-        } else {
-            self.iter.nth(self.step)
-        }
+        // branchless step size selection.
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:39]: Is branchless really better than letting the branch predictor run here?

Also, just using an `if` will give you a `select` in LLVM: <https://rust.godbolt.org/z/vGcKjrdc3>

I think the only reason you're getting something different from the `if` with this is that LLVM isn't smart enough to realize that you wrote a `select` here, because it ignores us telling it that the `load` of the bool will always be zero or one.

So I'm inclined to say that you also found a bug here worth filing on LLVM.

I'm not convinced that the mask hack is worth it here, though.

Can it not just be the simpler
```
let step_size = if self.first_take { 0 } else { self.step };
self.first_take = false;
self.iter.nth(step_size);
```
to get all the "only one instantiation" goodness, but still probably let LLVM understand what's going on and lower the `select` as it thinks is best?

File: library/core/src/iter/adapters/step_by.rs [hunk 2]
```diff
@@ -233,3 +225,96 @@ where
...
+
+macro_rules! spec_int_ranges {
+    ($($t:ty)*) => ($(
+        impl SpecRange for StepBy<Range<$t>> {
+            #[inline]
+            fn spec_fold<Acc, F>(self, mut acc: Acc, mut f: F) -> Acc
+            where
+                F: FnMut(Acc, Self::Item) -> Acc
+            {
+                use crate::convert::TryFrom;
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:267]: Isn't `core` in 2021?  Why is this needed?

File: library/core/src/iter/adapters/step_by.rs [hunk 3]
```diff
@@ -233,3 +225,96 @@ where
...
+                    // has been partially advanced
+
+                    // Don't care about overshooting the range end since it'll be dropped at the end
+                    // of the function
+                    range.start = range.start.saturating_add(<$t>::try_from(self.step).unwrap_or(<$t>::MAX));
+                }
+
+                // if a step size larger than the type has been specified fall back to
+                // t::MAX which will never yield more than 1 item.
+                let step = <$t>::try_from(self.step + 1).unwrap_or(<$t>::MAX);
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:287]: suggestion: comment that `step + 1` cannot overflow since we subtracted it when creating the iterator.

File: library/core/src/iter/adapters/step_by.rs [hunk 4]
```diff
@@ -233,3 +225,96 @@ where
...
+                    i += step;
+                }
+
+                acc
+            }
+        }
+    )*)
+}
+
+#[cfg(target_pointer_width = "64")]
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:313]: Pondering: if the specialization is only ever used when the range fits in `usize`, might it make sense to convert to `usize` always, do the looping in that -- it's presumably a good type for the machine anyway -- and just cast to smaller when calling `f`?

File: library/core/src/iter/adapters/step_by.rs [hunk 5]
```diff
@@ -233,3 +225,96 @@ where
...
+            where
+                F: FnMut(Acc, Self::Item) -> Acc
+            {
+                use crate::convert::TryFrom;
+
+                let mut range = self.iter;
+
+                if !self.first_take {
+                    // `next()` implements `increment-by-step, get-and-increment-by-one`, leaving
+                    // range.start set to one element after the previously returned one.
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:273]: It doesn't have to -- https://github.com/rust-lang/rust/pull/52193/files FCP'd that it's allowed to have `start` be the next thing to return after a `next()`.

Would it be worth using that freedom here?

**Review by scottmcm — COMMENTED** [review]
Sorry for taking a while to get to this.  I left a bunch of disjointed comments; let me know your thoughts.

@rustbot author

File: library/core/src/iter/adapters/step_by.rs (hunk 5 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:273]: That would mean changing or specializing all forward iteration methods to have that behavior, including `size_hint`.

With `advance_by` we can now do that in a reasonable manner but that's a bigger change and one that impacts observable behavior (even if allowed) so that should be done separately imo.

File: library/core/src/iter/adapters/step_by.rs (hunk 1 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:39]: Fun....

```
branchless:
    iter::bench_range_step_by       38.00ns/iter +/- 0.00ns
if-else:
    iter::bench_range_step_by       69.00ns/iter +/- 1.00ns
```

But that's with `-Ctarget-cpu=native`!
(and idk why it got even faster than what's in the PR description)

With x86-64 baseline I get:

```
branchless:
    iter::bench_range_step_by       39.00ns/iter +/- 0.00ns
if-else:
    iter::bench_range_step_by       39.00ns/iter +/- 0.00ns
```

😮‍💨

File: library/core/src/iter/adapters/step_by.rs (hunk 2 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:267]:
```
error[E0599]: no function or associated item named `try_from` found for type `usize` in the current scope
   --> library/core/src/iter/adapters/step_by.rs:290:39
    |
260 | / macro_rules! spec_int_ranges {
261 | |     ($($t:ty)*) => ($(
262 | |         impl SpecRange for StepBy<Range<$t>> {
263 | |             #[inline]
...   |
290 | |                 let inner_len = <$t>::try_from(range.size_hint().0).unwrap_or(<$t>::MAX);
    | |                                       ^^^^^^^^ function or associated item not found in `usize`
...   |
311 | |     )*)
312 | | }
    | |_- in this expansion of `spec_int_ranges!`
...
315 |   spec_int_ranges!(u8 u16 u32 u64 usize);
    |   -------------------------------------- in this macro invocation
    |
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
1   + use crate::convert::TryFrom;
    |
```

Maybe due to macros?

File: library/core/src/iter/adapters/step_by.rs (hunk 4 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:313]: I'm having some bistability issues with the benchmarks, but at least when targeting x86-64-v3 then using `$t`  different assembly compared to `usize`.


u16 as u16, this also results in better walltime results:

```
       │       vmovd        %ecx,%xmm0                                                                                                                                                                                                                                                                                      ▒
  0.49 │       vpbroadcastw %xmm0,%ymm0                                                                                                                                                                                                                                                                                     ▒
       │       vpaddw       -0x16a5e1(%rip),%ymm0,%ymm0        # e9a0 <str.1-0x660>                                                                                                                                                                                                                                         ▒
       │       mov          %r8d,%esi                                                                                                                                                                                                                                                                                       ▒
       │       shl          $0x6,%esi                                                                                                                                                                                                                                                                                       ▒
       │       lea          (%rcx,%rsi,1),%r10d                                                                                                                                                                                                                                                                             ▒
       │       vxorpd       %xmm1,%xmm1,%xmm1                                                                                                                                                                                                                                                                               ▒
       │       mov          %r8d,%r9d                                                                                                                                                                                                                                                                                       ▒
  0.97 │       vpxor        %xmm2,%xmm2,%xmm2                                                                                                                                                                                                                                                                               ▒
       │       vpxor        %xmm3,%xmm3,%xmm3                                                                                                                                                                                                                                                                               ▒
       │       vpxor        %xmm4,%xmm4,%xmm4                                                                                                                                                                                                                                                                               ▒
       │       vmovdqa      str.1-0xae0,%ymm10                                                                                                                                                                                                                                                                              ▒
       │       vmovdqa      str.1-0x15e0,%ymm7                                                                                                                                                                                                                                                                              ▒
       │       vmovdqa      str.1-0x400,%ymm11                                                                                                                                                                                                                                                                              ▒
       │       cs           nopw 0x0(%rax,%rax,1)                                                                                                                                                                                                                                                                           ▒
  4.90 │b40:┌─→vpaddw       %ymm0,%ymm14,%ymm5                                                                                                                                                                                                                                                                              ▒
  0.97 │    │  vpaddw       %ymm0,%ymm15,%ymm6                                                                                                                                                                                                                                                                              ▒
       │    │  vpxor        %ymm5,%ymm6,%ymm5                                                                                                                                                                                                                                                                               ▒
 18.53 │    │  vpaddw       %ymm0,%ymm12,%ymm6                                                                                                                                                                                                                                                                              ▒
  0.49 │    │  vpaddw       %ymm2,%ymm5,%ymm2                                                                                                                                                                                                                                                                               ▒
  0.49 │    │  vpaddw       %ymm0,%ymm10,%ymm5                                                                                                                                                                                                                                                                              ▒
       │    │  vpxor        %ymm6,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
  9.27 │    │  vpaddw       %ymm0,%ymm13,%ymm6                                                                                                                                                                                                                                                                              ▒
  2.93 │    │  vpaddw       %ymm3,%ymm5,%ymm3                                                                                                                                                                                                                                                                               ▒
  1.95 │    │  vpaddw       %ymm7,%ymm0,%ymm5                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpxor        %ymm6,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
 11.18 │    │  vpcmpeqd     %ymm6,%ymm6,%ymm6                                                                                                                                                                                                                                                                               ▒
       │    │  vpaddw       %ymm4,%ymm5,%ymm4                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpaddw       %ymm6,%ymm0,%ymm5                                                                                                                                                                                                                                                                               ▒
  1.94 │    │  vpxor        %ymm0,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
 10.70 │    │  vpaddw       %ymm1,%ymm5,%ymm1                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpaddw       %ymm0,%ymm11,%ymm0                                                                                                                                                                                                                                                                              ▒
  0.49 │    │  add          $0xffffffc0,%r9d                                                                                                                                                                                                                                                                                ▒
       │    └──jne          b40                                                                                                                                                                                                                                                                                             ▒
       │       vpaddw       %ymm1,%ymm2,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vpaddw       %ymm0,%ymm3,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vpaddw       %ymm0,%ymm4,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vextracti128 $0x1,%ymm0,%xmm1                                                                                                                                                                                                                                                                                ▒
  2.96 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  1.94 │       vpshufd      $0xee,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  2.45 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  0.50 │       vpshufd      $0x55,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.46 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vpsrld       $0x10,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ◆
  3.40 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vmovd        %xmm0,%r9d                                                                                                                                                                                                                                                                                      ▒
  0.49 │       cmp          %edi,%r8d                                                                                                                                                                                                                                                                                       ▒
       │     ↑ je           a70                                                                                                                                                                                                                                                                                             ▒
       │       test         $0x38,%dil                                                  
```

u16 as usize:

```
       │       vmovd        %edx,%xmm0                                                                                                                                                                                                                                                                                      ▒
       │       vpbroadcastw %xmm0,%ymm0                                                                                                                                                                                                                                                                                     ▒
       │       vpaddw       %ymm6,%ymm0,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       mov          %r9,%rsi                                                                                                                                                                                                                                                                                        ▒
       │       shl          $0x6,%rsi                                                                                                                                                                                                                                                                                       ▒
  0.73 │       lea          (%rsi,%rdx,1),%r10                                                                                                                                                                                                                                                                              ▒
       │       vxorpd       %xmm1,%xmm1,%xmm1                                                                                                                                                                                                                                                                               ▒
       │       mov          %r9,%r8                                                                                                                                                                                                                                                                                         ▒
       │       nop                                                                                                                                                                                                                                                                                                          ▒
  3.19 │a20:┌─→vpaddw       %ymm7,%ymm0,%ymm2                                                                                                                                                                                                                                                                               ▒
  2.48 │    │  vpxor        %ymm0,%ymm2,%ymm2                                                                                                                                                                                                                                                                               ▒
 51.25 │    │  vpaddw       %ymm1,%ymm2,%ymm1                                                                                                                                                                                                                                                                               ▒
  4.54 │    │  vpaddw       %ymm0,%ymm8,%ymm0                                                                                                                                                                                                                                                                               ▒
  9.94 │    │  add          $0xfffffffffffffff0,%r8                                                                                                                                                                                                                                                                         ▒
       │    └──jne          a20                                                                                                                                                                                                                                                                                             ▒
  0.02 │       vextracti128 $0x1,%ymm1,%xmm0                                                                                                                                                                                                                                                                                ▒
  1.11 │       vpaddw       %xmm0,%xmm1,%xmm0                                                                                                                                                                                                                                                                               ▒
  2.16 │       vpshufd      $0xee,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  2.26 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  0.12 │       vpshufd      $0x55,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.40 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vpsrld       $0x10,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.55 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vmovd        %xmm0,%r8d                                                                                                                                                                                                                                                                                      ▒
  0.75 │       cmp          %r9,%rdi                                                                                                                                                                                                                                                                                        ▒
       │     ↑ je           900                                                                                                                                                                                                                                                                                             ▒
       │       test         $0x8,%dil                               
```

File: library/core/src/iter/adapters/step_by.rs (hunk 5 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs]: Ok, turns out this was necessary anyway. I have no specialized all methods so we can operate differently on `Range<{integer}>`.

File: library/core/src/iter/adapters/step_by.rs [hunk 6]
```diff
@@ -32,16 +42,154 @@ where
...
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.spec_size_hint()
+    }
+
+    #[inline]
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.spec_nth(n)
+    }
+
+    fn try_fold<Acc, F, R>(&mut self, acc: Acc, f: F) -> R
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:58]: Should this also have inline?
```suggestion
    #[inline]
    fn try_fold<Acc, F, R>(&mut self, acc: Acc, f: F) -> R
```

File: library/core/src/iter/adapters/step_by.rs [hunk 7]
```diff
@@ -230,6 +360,163 @@ where
...
+                    // will not be used
+                    self.iter.start = val.wrapping_add(step);
+                    self.iter.end = remaining - 1;
+                    Some(val)
+                } else {
+                    None
+                }
+            }
+
+            fn spec_size_hint(&self) -> (usize, Option<usize>) {
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:400]: This is a concrete impl, so I think it needs to be explicitly `#[inline]` to be inlined outside the crate, no?

File: library/core/src/iter/adapters/step_by.rs [hunk 8]
```diff
@@ -13,12 +18,17 @@ use crate::{intrinsics, iter::from_fn, ops::Try};
 pub struct StepBy<I> {
     iter: I,
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:19]: Could you please put a comment on this to note that `setup` has always been run on this `iter`, as a type invariant?  `new` does that, but I never noticed that on my first read through here, and was really confused below.

File: library/core/src/iter/adapters/step_by.rs [hunk 9]
```diff
@@ -230,6 +360,163 @@ where
     }
 }
 
-// StepBy can only make the iterator shorter, so the len will still fit.
-#[stable(feature = "iterator_step_by", since = "1.28.0")]
-impl<I> ExactSizeIterator for StepBy<I> where I: ExactSizeIterator {}
+macro_rules! spec_int_ranges {
```
Comment (scottmcm) [inline library/core/src/iter/adapters/step_by.rs:363]: I think it would be nice to have a comment on this describing the approach.

A first stab:
```suggestion
// For these implementations, `SpecRangeSetup` calculates the number
// of iterations that will be needed and stores that in `iter.end`.
//
// The various iterator implementations then rely on that to not need
// overflow checking, letting loops just be counted instead.
macro_rules! spec_int_ranges {
```

**Comment by scottmcm** [comment]
I did have a bunch of requests, but overall this looks good.  `r=me` with some comment tweaks and maybe attribute updates.

File: library/core/src/iter/adapters/step_by.rs (hunk 6 above)
Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:58]: This one didn't have it before. But I'll add it to the concrete impls.

File: library/core/src/iter/adapters/step_by.rs [hunk 10]
```diff
@@ -32,16 +45,174 @@ where
...
+///
+/// # Correctness
+///
+/// Technically this is safe to implement (look ma, no unsafe!), but in reality
+/// a lot of unsafe code relies on ranges over integers being correct.
+///
+/// For correctness *all* public StepBy methods must be specialized
+/// because `setup` drastically alters the meaning of the struct fields so that mixing
+/// different implementations would lead to incorrect results.
+trait StepByImpl<I> {
```
Comment (the8472) [inline library/core/src/iter/adapters/step_by.rs:152]: Any opinion on making this an unsafe trait? As the comment says, it technically isn't, but getting it right is tricky and we rely on it for unchecked indexing in some places I think.

Comment (scottmcm) [reply library/core/src/iter/adapters/step_by.rs:152]: Hmm, are any of these things `TrustedLen`?  If so, `spec_size_hint` needs to be accurate to meet `TrustedLen`, so if `StepBy` is delegating that to the other trait, the other trait plausibly does need to be `unsafe trait StepByImpl`, yeah.

Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:152]: Currently it doesn't. But might as well add it at this point, it makes the decision simpler.

Comment (the8472) [reply library/core/src/iter/adapters/step_by.rs:152]: On the other hand `Zip`'s internal `ZipImpl` isn't marked as unsafe even though it has similar properties.

**Comment by the8472** [comment]
the mips runner is flaky. see https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/MIPS.20failure.20on.20CI/near/368404515

@bors retry
